### DISCUSSION TITLE
Using eza in large directories

### DISCUSSION THREAD

**Discussion by apcamargo** [eza-community/eza#816] [Q&A, Howto, Help]
A common part of my workflow involves executing `ls` in large directories (specifically, `ls -l large_dir`), which prevents me from setting `ls` as an alias for `eza`, since `eza` is significantly slower in large directories compared to `ls`. This isn't a major issue, and I understand that it's probably not what `eza` is mainly for. Still, I'm curious if there's any combination of parameters that could make `eza` run faster when fancy stuff, like Git details, are not really necessary.

**Comment by cafkafk** [comment]
I just ran eza in a folder with 26218 files and it took an average of 104 miliseconds

```
  Time (mean ± σ):     103.9 ms ±   1.9 ms    [User: 39.5 ms, System: 64.2 ms]
  Range (min … max):   100.4 ms … 107.9 ms    28 runs
```

I'm assuming you're running with additional flags. These are likely the reason that your seeing slow speeds. Especially flags like `--git-repos` will slow down the speed on large dirs, which is why `--git-repos-no-status` exists, for instance.

**Reply by apcamargo** [reply]
The only flag I'm using is `-1` because it's important for me to have one file per line. So nothing really complex.

Here's my benchmark in a directory with ~50k files. The difference gets bigger when the number of files increases.

```
Benchmark 1: ls -1 dir | wc -l
  Time (mean ± σ):     807.9 ms ± 219.8 ms    [User: 102.2 ms, System: 20.4 ms]
  Range (min … max):   597.6 ms … 1036.0 ms    3 runs

Benchmark 2: eza -1 dir | wc -l
  Time (mean ± σ):      3.541 s ±  0.422 s    [User: 0.203 s, System: 0.654 s]
  Range (min … max):    3.074 s …  3.895 s    3 runs

Summary
  ls -1 dir | wc -l ran
    4.38 ± 1.30 times faster than eza -1 UHGV | wc -l
```

**Reply by cafkafk** [reply]
Again, I fail to reproduce your speeds locally, here in a folder with 50k files

```
✦ ➜ eza -1 | wc -l
50000
~/test 
✦ ➜ hyperfine 'eza -1'
Benchmark 1: eza -1
  Time (mean ± σ):     157.1 ms ±   4.2 ms    [User: 71.7 ms, System: 85.2 ms]
  Range (min … max):   151.6 ms … 167.0 ms    19 runs
 ```
 
 This is still slower than `ls`, no doubt
 
 ```
 ✦ ➜ hyperfine 'ls -1'
Benchmark 1: ls -1
  Time (mean ± σ):      36.5 ms ±   2.3 ms    [User: 25.4 ms, System: 10.9 ms]
  Range (min … max):    33.0 ms …  45.7 ms    78 runs
  ```
  
  What kind of machine are you running on, what version of eza are you using? Did you compile it yourself (in release mode)?

**Reply by apcamargo** [reply]
The ratio between `eza` and `ls` in your benchmark is similar to the one I got, so I think we are seeing pretty much the same different. Both `ls` and `eza` were slower on my end, but this might just because I'm using a cluster with a lot of users doing I/O operations or because the SSDs might be slower than yours. The version of `eza` I'm using is `v0.18.0`.

Benchmarking on five directories with a total of ~102k files makes the difference more pronounced:

```
Benchmark 1: ls -1 dir1 dir2 dir3 dir4 dir5 | wc -l
  Time (mean ± σ):     557.1 ms ± 122.4 ms    [User: 148.5 ms, System: 25.8 ms]
  Range (min … max):   429.0 ms … 673.0 ms    3 runs

Benchmark 2: eza -1 dir1 dir2 dir3 dir4 dir5 | wc -l
  Time (mean ± σ):     95.422 s ±  8.538 s    [User: 0.402 s, System: 1.934 s]
  Range (min … max):   85.575 s … 100.778 s    3 runs

  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet system without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.

Summary
  ls -1 dir1 dir2 dir3 dir4 dir5 | wc -l ran
  171.29 ± 40.65 times faster than eza -1 dir1 dir2 dir3 dir4 dir5 | wc -l
```

**Reply by cafkafk** [reply]
Okay I see, while that's hardly an ideal place to test runtime, I do think there is space for improvement in general. In the past we've talked about taking some of the lessons from https://github.com/saethlin/fls and using them in eza, but I don't think any of the currently active devs are viewing this as a priority.

**Reply by apcamargo** [reply]
Yes, I agree it is not the best environment, but the difference is substantial enough to make it clear that there's a big difference in performance. But I do understand this is probably not the usual use case for `eza`.

**Reply by daviessm** [reply]
@apcamargo it looks like @tertsdiepraam has just created #823 which claims speed improvements, would you like to check their branch out and see if it improves things on your setup?

**Reply by tertsdiepraam** [reply]
My PR was about the details view, so I think it's unrelated to `-1`. Sounds like a fun challenge though 😄

**Reply by tertsdiepraam** [reply]
I think at least one issue is that `eza` always does a `statx` call on every single file, even though it doesn't need to. Even on a oneline format, it first collects all the metadata it might need and then passes the `File` struct (not the `std::file::File` but the one defined in `src::fs::File`) to the renderer.

There's several ways to fix this:
- Split the renderers into renderers that do and do not require more metadata. This can get quite tricky, because even the details view might not need the metadata if all the columns that require metadata are disabled. It might still be a pretty good middle ground between complexity and speed though.
- Put the metadata in a `OnceCell` meaning that it's only requested when necessary. This is what uutils does, but it's not super pretty, but it works quite well still. I introduced that in this PR (but it's a couple years old by now, so I probably would have written it better today 😄): https://github.com/uutils/coreutils/pull/2104

**Reply by daviessm** [reply]
> There's several ways to fix this:

Sounds like you're volunteering 😄 But thanks!

**Reply by tertsdiepraam** [reply]
Actually, this strategy is used already for the extended attributes and absolute path.

**Reply by tertsdiepraam** [reply]
> Sounds like you're volunteering

Happy to if it turns out to be a relatively small fix!

**Reply by apcamargo** [reply]
Should I open an issue to track this?

**Reply by tertsdiepraam** [reply]
Good idea! I started this PR once that you could also link to: https://github.com/eza-community/eza/pull/833. It was more of an exploration than than a real PR.

**Reply by apcamargo** [reply]
Done (https://github.com/eza-community/eza/issues/922). Sorry that I took so long to do this
